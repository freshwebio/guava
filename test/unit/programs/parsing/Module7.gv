module Order exposing { total }

function total(priceStr, taxPerc, extractPriceInPence) :: String -> String -> (String -> String -> Integer) -> Float {
  extractPriceInPence(priceStr, taxPerc) / 100
}