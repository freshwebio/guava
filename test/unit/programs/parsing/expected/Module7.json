{
  "description": "should parse a program with a function that returns a floating point with arguments and an inline type definition with mutliple parameters and an output type",
  "module": {
    "name": "Order",
    "exposes": {
      "exposed": ["total"]
    },
    "moduleStatements": {
      "statementList": [{
        "kind": "function",
        "fncStatement": {
          "kind": "functionWithInlineTypeDef",
          "functionName": "total",
          "defArgs": {
            "args": [{
              "kind": "defArgName",
              "argName": "priceStr"
            }, {
              "kind": "defArgName",
              "argName": "taxPerc"
            }, {
              "kind": "defArgName",
              "argName": "extractPriceInPence"
            }]
          },
          "functionTypeDefValueTypes": [{
            "kind": "valueTypeSingle",
            "valueTypeName": "String"
          }, {
            "kind": "valueTypeSingle",
            "valueTypeName": "String"
          }, {
            "kind": "valueTypeFunction",
            "functionValueTypes": [{
              "kind": "valueTypeSingle",
              "valueTypeName": "String"
            }, {
              "kind": "valueTypeSingle",
              "valueTypeName": "String"
            }, {
              "kind": "valueTypeSingle",
              "valueTypeName": "Integer"
            }]
          }, {
            "kind": "valueTypeSingle",
            "valueTypeName": "Float"
          }],
          "functionExpr": {
            "kind": "arithmeticExpr",
            "arithmeticExpr": {
              "kind": "aBinary",
              "arithmeticBinaryOperator": {
                "kind": "divide"
              },
              "arithmeticExpressionLeft": {
                "kind": "arithmeticFunctionCall",
                "calledArithmeticFunctionName": "extractPriceInPence",
                "arithmeticFuncCallArgs": {
                  "kind": "callArgs",
                  "callArgs": [{
                    "kind": "var",
                    "varName": "priceStr"
                  }, {
                    "kind": "var",
                    "varName": "taxPerc"
                  }]
                }
              },
              "arithmeticExpressionRight": {
                "kind": "integerConst",
                "intConstValue": 100
              }
            }
          }
        }
      }]
    }
  }
}